cpp/include/raft/core/mdbuffer.cuh

```diff
@@ -0,0 +1,999 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+#include <algorithm>
+#include <execution>
+#include <optional>
+#include <raft/core/cuda_support.hpp>
+#include <raft/core/detail/copy.hpp>
+#include <raft/core/device_container_policy.hpp>
+#include <raft/core/error.hpp>
+#include <raft/core/host_container_policy.hpp>
+#include <raft/core/host_device_accessor.hpp>
+#include <raft/core/logger.hpp>
+#include <raft/core/managed_container_policy.hpp>
+#include <raft/core/mdarray.hpp>
+#include <raft/core/mdspan.hpp>
+#include <raft/core/pinned_container_policy.hpp>
+#include <raft/core/stream_view.hpp>
+#include <raft/util/variant_utils.hpp>
+#include <type_traits>
+#include <utility>
+#include <variant>
+#ifndef RAFT_DISABLE_CUDA
+#include <raft/core/copy.cuh>
+#include <raft/util/cudart_utils.hpp>
+#include <thrust/device_ptr.h>
+#else
+#include <raft/core/copy.hpp>
+#endif
+
+namespace raft {
+
+/**
+ * @defgroup mdbuffer_apis multi-dimensional maybe-owning type
+ * @{
+ */
+
+/**
+ * @brief Retrieve a canonical index associated with a given memory type.
+ *
+ * For variants based on memory type, this index can be used to help keep a
+ * consistent ordering of the memory types in the variant.
+ */
+inline auto constexpr variant_index_from_memory_type(raft::memory_type mem_type)
+{
+  return static_cast<std::underlying_type_t<raft::memory_type>>(mem_type);
+}
+
+/**
+ * @brief Retrieve the memory type associated with a canonical index
+ */
+inline auto constexpr memory_type_from_variant_index(
+  std::underlying_type_t<raft::memory_type> index)
+{
+  return static_cast<raft::memory_type>(index);
+}
+
+/**
+ * @brief Retrieve a type from a variant based on a given memory type.
+ */
+template <raft::memory_type MemType, typename Variant>
+using alternate_from_mem_type =
+  std::variant_alternative_t<variant_index_from_memory_type(MemType) % std::variant_size_v<Variant>,
+                             Variant>;
+
+namespace detail {
+template <typename T, raft::memory_type MemType>
+using memory_type_to_default_policy = std::conditional_t<
+  MemType == raft::memory_type::host,
+  host_vector_policy<T>,
+  std::conditional_t<
+    MemType == raft::memory_type::device,
+    device_uvector_policy<T>,
+    std::conditional_t<
+      MemType == raft::memory_type::managed,
+      managed_uvector_policy<T>,
+      std::conditional_t<MemType == raft::memory_type::pinned, pinned_vector_policy<T>, void>>>>;
```

Comment:
This is totally subjective, but maybe partial instantiation here is a little bit more readable by avoiding nested conditional_t?
```suggestion
template <typename T, raft::memory_type MemType>
struct memory_type_to_default_policy {};
template <typename T>
struct memory_type_to_default_policy<T, raft::memory_type::host> {
    using type = typename raft::host_vector_policy<T>;
};
template <typename T>
struct memory_type_to_default_policy<T, raft::memory_type::device> {
    using type = typename raft::device_uvector_policy<T>;
};
template <typename T>
struct memory_type_to_default_policy<T, raft::memory_type::managed> {
    using type = typename raft::managed_uvector_policy<T>;
};
template <typename T>
struct memory_type_to_default_policy<T, raft::memory_type::pinned> {
    using type = typename raft::pinned_vector_policy<T>;
};

template <typename T, raft::memory_type MemType>
using memory_type_to_default_policy_t = typename memory_type_to_default_policy<T, MemType>::type;
```

Reply:
Sure! I don't mind it